### ISSUE TITLE
Bootstrap-Vue V2.0.0-rc.12 Tracker

### ISSUE THREAD

**Issue by tmorehouse** [bootstrap-vue/bootstrap-vue#2173]
This issue is for tracking progress of the next release of Bootstrap-Vue.

We will be updating this list as PRs are added/merged.

_If you have bug reports or feature requests, please open a [**new issue**](https://github.com/bootstrap-vue/bootstrap-vue/issues/new) rather than comment here._

<hr>

### New Features
- [x] Use Bootstrap V4.3 SCSS
- [x] `b-modal` Allow multiple modal's opened at once (stacked) (PR #2167)
- [x] `b-table` Add support for selectable rows (PR #2260, closes #1790)
- [x] `b-form-checkbox` and `b-form-checkbox-group` Bootstrap V4.2 switch style checkboxes (PR #2293)
- [x] `b-spinner` New Bootstrap V4.2 component (PR #2306)
- [x] Migrate CSS to SCSS for custom styles and theming support (PR #2221) _Big thanks to @jackmu95!_

### Enhancements
- [x] `b-alert` add fade prop and fade animation CSS  (PR #1785, #2205)
- [x] `b-card` new helper sub-components `b-card-title`, `b-card-sub-title` and `b-card-text` (PR #2375)
- [x] `b-carousel` added support for swipe left/right on touch screens (PR #2409, closes #2358)
- [x] `b-dropdown` add dropleft and drop right support (PR #2117)
- [x] `b-dropdown` make show and hide events cancelable (PR #1807)
- [x] `b-dropdown` added support for link/router-link and 'variant' in split button mode (PR  #2301, Closes #1960)
- [x] `b-dropdown` added support for form controls and free flow text (PR #2434)
- [x] `b-dropdown` and `b-navbar-item-dropdown` use provide and inject for sub items to control closing dropdown (PR #2431)
- [x] `b-form-input` implement bootstrap v4.1 custom range input (PR #2122, #2120)
- [x] `b-form-texarea` incorporate optional formatter (PR #2111)
- [x] `b-form-invlaid-feedback` and `b-form-valid-feedback` add support for tooltip style (PR #2188)
- [x] `b-form-radio` and `b-form-checkbox` switch to use provide and inject for better detection of grouping (PR #2241)
- [x] `b-form-group` support multiple breakpoint label column sizes and alignment (PR #2258, Closes #2230, #1973)
- [x] `b-list-group` add support responsive horizontal list-groups (PR #2536)
- [x] `b-modal` new v4.2 `xl` size
- [x] `b-modal` add support for scrollable modal content (PR #2535)
- [x] `b-progress` & `b-progress-bar` use provide and inject for inter component communication (PR #2540)
- [x] `b-table` use SCSS and bootstrap variables for defining stacked table breakpoints
- [x] `b-table` improved codebase for filtering, sorting, and pagination (PR #2149, fixes #1517, #1989)
- [x] `b-table` add contextmenu event on rows (PR #2064)
- [x] `b-table` add row uncovered event (PR #1874)
- [x] `b-table` allow sorting on nested object properties (PR #1868)
- [x] `b-table` support row arbitrary classes (PR #1797, fixes #1705)
- [x] `b-table` introduce option `table-busy` slot for use loading message and/or spinners, etc (PR #2196, Closes #1859)
- [x] `b-table` no longer need to add `.stop` modifier on `click` events inside data or head scoped slots to prevent `head-clicked` or `row-clicked` emitting (PR #2214)
- [x] `b-table` add `borderless` style prop  (PR #2300)
- [x] `b-table` add support for transitions on `<tbody>` element (PR #2450, Closes #1821)
- [x] `b-tabs` active tabs dynamically via `active` prop and public methods (PR #2442)
- [x] `b-popover` include directive in component plugin (PR #2115)
- [x] `b-tooltip` include directive in component plugin (PR #2116)

### Fixes
- [x] security: strip script tags from html injected into elements (PR #2134, fixes #1974, #1665)
- [x] auto id: ensure reactivity of auto generated id and apply only after mount() (PR #2161, fixes #1978)
- [x] `b-button`, `b-dropdown`, and `b-link`: fix various interrelated issues and aria (PR #2159, fixes #1814, #1817)
- [x] `b-dropdown` support links (and other tags) as trigger elements (PR #1929)
- [x] `b-dropdown` better `no-caret` option using SCSS (PR #2136, fixes #1473)
- [x] `b-dropdown-item` fix click event handling bug in Safari (PR #2252)
- [x] `b-dropdown-item-button` Add support for `active` state (PR #2212)
- [x] `b-card` and sub component render issues with undefined slots (PR #2125, fixes #2062)
- [x] `b-card` add support for left and right image placement (PR #1981)
- [x] all form controls: allow attribute `form` to be applied to custom inputs (PR #2172, fixes #2154)
- [x] all form controls: fix state prop when set to empty string (PR #2167, fixes #2166)
- [x] `b-form-input` bugfixes (PR #2100, #2204)
- [x] `b-form-input` attach native event listeners (PR #1873)
- [x] `b-form-file` allow input to be reset by clearing v-model (PR #2170)
- [x] `b-form-file` customize  browse button text (PR #2168, fixes #2143)
- [x] `b-form-file` fix faulty drag and drop event handlers (PR #2169)
- [x] `b-form-select` fire change event after input (PR #2207, Fixes #1772)
- [x] `b-form-select` remove interim classes needed for bootstrap v4.0.x (PR #2265, Closes #1896)
- [x] `b-form-textarea` bugfixes (PR #2111)
- [x] `b-form-textarea` attach native listeners (PR #2111)
- [x] `b-input-group` styling fix for dropdowns, radios, and checks (PR #2118, fixes #2114, #1560)
- [x] `b-list-group-item` default button type to button for button action items (PR #2194, Fixes #2192)
- [x] `b-modal` add content class prop (PR #1961, fixes #1950)
- [x] `b-modal` optimize reset Scrollbar routine (PR #1837, fixes #1831)
- [x] `b-modal` handle enforce focus when modal in stacked (multiple opened) mode (PR #2211, fixes #2175)
- [x] `b-modal` ability to disable stacking mode (PR #2259)
- [x] `b-modal` handle edge cases where model programatically show and hidden in rapid succession (PR #2270, Fixes #2236)
- [x] `b-pagination` & `b-pagination-nav` add slots for bookend buttons and ellipsis (PR #1980)
- [x] `b-pagination` & `b-pagination-nav` v-model active class fix + keypress click fix (PR #2299, Fixes #1985, #1629, Closes: #1891)
- [x] `b-table` return empty string for undefined cell data (PR #2139, fixes #1502)
- [x] `b-table` fix stacked breakpoints to match latest Bootstrap V4.1 css (PR #1933)
- [x] `b-table` preserve attributes `aria-rowcount` and `aria-describedby` if provided  (PR #2195, Fixes #1801)
- [x] `b-table` Emit v-model input event only when computedItems changes (PR #2254, Closes #2231)
- [x] `b-table` handle edge cases for clearing selected rows (PR #2267)
- [x] `b-tabs` Better reactivity of tab title slots (PR #2442)
 
### Documentation
- [x] enhanced component reference for sub components (PR #2132)
- [x] fix live example (v-play) console errors during live edits (PR #2176)
- [x] fix playground hangs on template & code errors (PR #2177, #2193, #2197, fixes #1843)
- [x] Use babel-standalone` in playground and v-play to allow most es6 examples to work in older browsers (PR #2286, #2294)
- [x] playground now supports render functions
- [x] document vue-loader v15 changes
- [x] vee-validate examples (PR #1938, #1889 , fixes #1270)
- [x] Vue-cli v3 usage documentation (PR #2323, Closes #2096, #1946)
- [x] various documentation fixes and updates

### Unit Testing
- [x] Use `vue-test-utils` for component testing (on some components)

### New Features
- [ ] `b-toast` (and `b-toaster`) Bootstrap V4.2 toast component _(may wait for BS V4.2 to be released)_
- [ ] Use `portal-vue` for tooltips/popovers and modals, as well as toasts _(needs Portal-Vue 2.0 when available)_

### Enhancements
- [ ] `b-table` optional sorting and filtering of formatted and scoped slot item fields (when local sorting or local filtering only)
- [ ] `b-navbar` `b-collapse` `b-bavbar-toggle` use provide and inject to auto-associate toggle to collapse
- [ ] `b-modal` scoped/parameterized modals
- [ ] `b-popover` and `b-tooltip` add support for custom class on root tip/pop element _(wait on revamp of popover/tooltip/modal/toast)_

### Unit testing
- [ ] Use `vue-test-utils` for component testing

### Other
- [ ] Webpack template updates to use latest Vue/Boostrap-Vue/Bootstrap

**Comment by tmorehouse** [comment]
We're looking for some help with adding documentation re Vue CLI v3

**Comment by robcresswell** [comment]
I'll take it. Used the CLI a fair bit so should have enough know-how.

**Comment by noahbornstein** [comment]
Great work all - love this project and really appreciate it! Any rough estimates on next release date? (days? weeks? months?) Thanks!

**Comment by tmorehouse** [comment]
Next release will be in more than 1 second into the future.

Hoping we can get something out this week.  We may not have all the kinks we want ironed out, but at least there will be a lot of fixes.

**Comment by ghenry** [comment]
Any chance of this in the release? https://github.com/bootstrap-vue/bootstrap-vue/issues/1564

**Comment by tmorehouse** [comment]
Probably not in this coming release (which we are hoping to have out this week).

It wouldn't be something we could easily add to form-group, as this feature relies on specific markup (where the label is after the input), and doesn't handle input sizing based on their custom CSS (which is not part of bootstrap V4.1)

But  you can definitely wrap a regular input in the required markup and classes, and make a re-usable custom component.

**Comment by TitanFighter** [comment]
Hi @robcresswell if you have not started yet preparation of `Vue-cli v3 usage documentation`, I can take it, in order to make a new release with updated info.

**Comment by tmorehouse** [comment]
Is that from importing Bootstrap-Vue via the /es/  modules directory?  It should all be pre-transpiled

**Comment by TitanFighter** [comment]
It is not imported, there is no any`import xyz from "xyz"` at the top of `vue.config.js`. We simply need to add a string or regex to `transpileDependencies`.

1. From the [Guide](https://cli.vuejs.org/guide/browser-compatibility.html#usebuiltins-usage):

> If one of your dependencies need polyfills, you have a few options:
> 
> If the dependency is written in an ES version that your target environments do not support: Add that dependency to the transpileDependencies option in vue.config.js. This would enable both syntax transforms and usage-based polyfill detection for that dependency.

2. From the [Reference](https://cli.vuejs.org/config/#transpiledependencies)

> **transpileDependencies**
> Type: Array<string | RegExp>
> 
> Default: []
> 
> By default babel-loader ignores all files inside node_modules. If you want to explicitly transpile a dependency with Babel, you can list it in this option.

____

Regarding:
> It should all be pre-transpiled

... `bootsrap-vue` does not work with iOS 9 from the box.

From [Browser Support -> JS](https://bootstrap-vue.js.org/docs):
> BootstrapVue is written in Vue! So this is up to your project and bundler which browsers are supported. If you want to support older IE, Android and IOS devices, you may want to use Babel Polyfill.

I do not know for what target browsers `bootstrap-vue` is pre-transpiled, but it seems that the iOS 9 is out of the that range.

**Comment by tmorehouse** [comment]
In package.json we point `module` (and `js:next`) to the pre transpiled `/es/*` build

It might be because of our bablerc/browserlistrc files may not be targeting iOS9 properly.

Also note that there are two sub modules that are considered not "pure" which add in a couple of poly fills that could get filtered out (notably utils/object and utils/array)

Was going to add these exceptions to package.json to let webpack/etc know they have side effects.

**Comment by TitanFighter** [comment]
Just changed `import BootstrapVue from 'bootstrap-vue'` to `import BootstrapVue from 'bootstrap-vue/es'` + removed `transpileDependencies: ['bootstrap-vue']` and checked on iOS 9 -> it works :) One more thing to mention in the docs :)

**Comment by tmorehouse** [comment]
Interesting. It _should_ auto pick the `es/` modules based on our package.json file entries

**Comment by TitanFighter** [comment]
Just double checked without `/es` and without `transpileDependencies` and it works... Need more time to find out why previous attempts were unsuccessful.

**Comment by tmorehouse** [comment]
Are you using rc.11 or the dev branch for testing?

**Comment by TitanFighter** [comment]
rc.11

**Comment by tmorehouse** [comment]
Ok.. is it possible to check against the latest dev branch?

**Comment by TitanFighter** [comment]
For some reason I can not replicate error on iOS 9 using rc.11. Yes, I can check the latest dev branch. In about 1 hour I will come back with a report.

**Comment by tmorehouse** [comment]
Dev branch has an updated `.browserlistrc` file compared to current master. So it may have better coverage for a few older browsers.

**Comment by tmorehouse** [comment]
@TitanFighter any luck with your recent findings/tests?

**Comment by krgaurav1992** [comment]
Can you please add support for raw html so that icons can be inserted with breadcrumbs ?

What I want is:
`{ text: '<span class="lnr lnr-home"></span>', to: '/' }`

**Comment by jsancho-gpl** [comment]
Hi people, thank you so much for your excellent work! Is there any estimation for the rc12 release date?

**Comment by pi0** [comment]
@jsancho-gpl today :angel:

**Comment by tmorehouse** [comment]
BoostrapVue v2.0.0-rc.12 has been released today!

**Comment by alexandernst** [comment]
:party_parrot:
